src/gt4py/next/iterator/transforms/trace_shifts.py

```diff
@@ -322,46 +327,48 @@ def fun(*args):
 
         return fun
 
-    def visit_StencilClosure(self, node: ir.StencilClosure):
-        tracers = []
-        for inp in node.inputs:
+    def initialize_context(self, inputs: Iterable[ir.Sym | ir.SymRef]) -> dict[str, Any]:
+        ctx: dict[str, Any] = {**_START_CTX}
+        for inp in inputs:
             self.shift_recorder.register_node(inp)
-            tracers.append(IteratorArgTracer(arg=inp, shift_recorder=self.shift_recorder))
-
-        result = self.visit(node.stencil, ctx=_START_CTX)(*tracers)
-        assert all(el is Sentinel.VALUE for el in _primitive_constituents(result))
-        return node
+            ctx[inp.id] = ArgTracer(arg=inp, shift_recorder=self.shift_recorder)
+        return ctx
 
     @classmethod
-    def apply(
-        cls, node: ir.StencilClosure | ir.FencilDefinition, *, inputs_only=True, save_to_annex=False
-    ) -> (
-        dict[int, set[tuple[ir.OffsetLiteral, ...]]] | dict[str, set[tuple[ir.OffsetLiteral, ...]]]
+    def trace_stencil(
+        cls, stencil: ir.Expr, *, num_args: Optional[int] = None, save_to_annex: bool = False
     ):
+        if isinstance(stencil, ir.Lambda):
+            assert num_args is None or num_args == len(stencil.params)
+            num_args = len(stencil.params)
+        if not isinstance(num_args, int):
+            raise ValueError("Stencil must be an 'itir.Lambda' or `num_args` is given.")
+        assert isinstance(num_args, int)
+
+        args = [im.ref(f"__arg{i}") for i in range(num_args)]
+
         old_recursionlimit = sys.getrecursionlimit()
         sys.setrecursionlimit(100000000)
 
         instance = cls()
-        instance.visit(node)
+        ctx = instance.initialize_context(args)
+        instance.visit(im.call(stencil)(*args), ctx=ctx)
```

Comment:
This pattern looks a bit weird as initialization is done in 2 steps. Maybe you can disentangle a bit more?

Reply:
I cleaned it up a bit, but still not great. I would leave it as is for now.